introspect_test.go

```diff
@@ -0,0 +1,190 @@
+package libp2p
+
+import (
+	"context"
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/websocket"
+	"github.com/libp2p/go-libp2p-core/introspect"
+	"github.com/libp2p/go-libp2p-core/metrics"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-introspection/introspection"
+	"github.com/stretchr/testify/require"
+	"net/url"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestIntrospector(t *testing.T) {
+	msg1 := []byte("1")
+	msg2 := []byte("12")
+	msg3 := []byte("111")
+	msg4 := []byte("0000")
+
+	iaddr := "0.0.0.0:9999"
```

Comment:
Tests should not listen on 0.0.0.0 (that's insecure) nor on a fixed port (it might be occupied). Instead, we should listen on 127.0.0.1:0, and allow the listener to allocate a port. `ListenAddrs()` should then return the _actual_ addresses, not the user-supplied ones.